src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -1297,17 +1301,14 @@ protected bool IsGeneratedCode(SyntaxTree tree)
 
             Debug.Assert(_lazyGeneratedCodeFilesMap != null);
 
-            lock (_lazyGeneratedCodeFilesMap)
+            bool isGenerated;
+            if (!_lazyGeneratedCodeFilesMap.TryGetValue(tree, out isGenerated))
             {
-                bool isGenerated;
-                if (!_lazyGeneratedCodeFilesMap.TryGetValue(tree, out isGenerated))
-                {
-                    isGenerated = _isGeneratedCode(tree, analyzerExecutor.CancellationToken);
```

Comment:
Was the primary issue here that we were invoking this within a lock? I am wondering if splitting this into 2 separate lock statements gives better performance then use of concurrent dictionary...

Reply:
:arrow_right: The primary issue here is the `TryGetValue` is placed inside the lock.